libvast/src/segment_store.cpp

```diff
@@ -373,8 +372,15 @@ void segment_store::inspect_status(caf::settings& xs,
 }
 
 caf::error segment_store::register_segments() {
-  for (auto filename : directory{segment_path()})
-    if (auto err = register_segment(filename))
+  std::error_code err{};
+  std::filesystem::directory_iterator dir{
+    std::filesystem::path{segment_path().str()}, err};
```

Comment:
```suggestion
  std::filesystem::directory_iterator dir{segment_path(), err};
```

I wondered: why not this? ☝️ 

Then I realized that you may want to do the upstream adjustment in a separate PR, given the PR title. Am I interpreting this correctly?

Reply:
In the longer term, we'd certainly like to change the return type of the `segment_path()` function to return a `std::filesystem::path`. I actually went down that approach first of converting all of `segment_store` to use `std::filesystem` types, but the call-def chain got a bit larger than what I wanted to tackle at that time due to calling into other functions from other files with path types, etc if I recall correctly.  It had the same issues with integration tests failing, but I think it's due to the non-existent directory like you pointed out.  That's why I split up the changeset into just this.

So for now, I've taken the approach of the absolute minimal requirement needed to remove the `vast::directory` dependency from `segment_store`.

After this PR lands, I think we can just remove the remaining includes for `directory.hpp` and the associated `directory` class, and `vast::directory` will be gone.

After `directory` is gone, I'd propose going through and trying to surgically convert uses of `vast::path` with their `std::filesystem::path` equivalents.  The hope would be to do this in a fashion that neither introduces a lot of path conversions (back and forth between `vast::path` and `std::filesystem::path` due to existing APIs) or introduces separate overloads in the APIs for `std::filesystem::path` and `vast::path`.

Reply:
Thanks for shedding light into your thought process.

It sounds like you already have a solid plan how to proceed with the refactoring. I like your proposal. I'm all for doing small PRs (which also helps in bisecting) and treading carefully.